### PR TITLE
Implementation of the the FSD-SINet family of models

### PR DESCRIPTION
**Pull request by palonso**
This PR addresses the implementation of the algorithm required to use the FSD-SINet models by @edufonseca.
It merges three branches where different features required to reproduce the mel-spectrogram parametrization were implemented (#1296, #1297, #1298) so if we decide to address all the changes here at once these PRs could be closed.

One of the limitations to reproduce the original implementation is that, by default, Essentia operates with floats instead of doubles. This has some impact in regression of mel-spectrograms (relative error ~1e-1) but it doesn't affect much the predictions by the model (relative error ~1e-4).

Another note is that the clip-wise audio normalization is only implemented in standard mode since in real-time applications it is not possible to access the entire signal beforehand.

### CODE REVIEW

File: src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  "memory exhausting for long files.\n"
+  "\n"
+  "The recommended pipeline is as follows::\n"
+  "\n"
+  "  MonoLoader(sampleRate=22050) >> TensorflowPredictFSDSINet\n"
+  "\n"
+  "Note: This algorithm does not make any check on the input model so it is "
+  "the user's responsibility to make sure it is a valid one.\n"
+  "\n"
+  "Note II: The FSD-SINet models were trained on normalized audio clips. "
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp:162]: Keeping simply "Note:" for any amount of notes maybe OK for a cleaner look.

File: src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+  "Clip-level normalization is only implemented in standard mode since in streaming there is no access to the entire audio clip. "
+  "In this case, it is the responsibility of the user to feed the algorithm with audio featuring a healthy dynamic range.\n\n"
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp:164]: It is not clear what is the "healthy" range that is expected. Can it be more concrete?

File: src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  "References:\n"
+  "  [1] Fonseca, E., Ferraro, A., & Serra, X. (2021). Improving sound event classification by increasing shift invariance in convolutional neural networks. arXiv preprint arXiv:2107.00623.\n"
+  "  [2] https://github.com/edufonseca/shift_sec"
+);
+// todo: refer also the github repository if it is updated with the source code https://github.com/edufonseca/shift_sec
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp:169]: Outdated comment?

File: src/algorithms/standard/windowing.cpp [hunk 4]
```diff
@@ -51,6 +51,7 @@ const char* Windowing::description = DOC("This algorithm applies windowing to an
 void Windowing::configure() {
   _normalized = parameter("normalized").toBool();
   _window.resize(parameter("size").toInt());
+  _constantsDecimals = parameter("constantsDecimals").toInt();
   _symmetric = parameter("symmetric").toBool();
   _constantsDecimals = parameter("constantsDecimals").toInt();
```
Comment (dbogdanov) [inline src/algorithms/standard/windowing.cpp:56]: Duplicate line leftover from merging conflicts.

File: src/algorithms/standard/windowing.h [hunk 5]
```diff
@@ -81,6 +81,7 @@ class Windowing : public Algorithm {
   bool _normalized;
   bool _splitPadding;
   bool _symmetric;
+  bool _splitPadding;
```
Comment (dbogdanov) [inline src/algorithms/standard/windowing.h:84]: Duplicate line leftover from merging conflicts.

File: src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+  vector<vector<Real> >& predictions = _predictions.get();
+  vector<Real> normalizedSignal;
+
+  if (_normalize) {
+    normalizedSignal = *signal;
+    normalizeFSDSINet(normalizedSignal);
+    signal = &normalizedSignal;
+  }
+
+  if (!signal->size()) {
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp:226]: Could be done before normalize

File: test/src/unittests/spectral/test_tensorflowinputfsdsinet.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            filedir(),
+            "tensorflowinputfsdsinet",
+            "dc_melspectrogram_fsdsinet.npy"
+        )
+        expected = numpy.load(expected_file)
+
+        dc = numpy.ones(self.sr, dtype="float32")
+        found = self.melspectrogram(dc)
+
+        # Essentia generates and additional frame because of its zero-padding strategy.
```
Comment (dbogdanov) [inline test/src/unittests/spectral/test_tensorflowinputfsdsinet.py:68]: "an additional frame"

**Review by dbogdanov — CHANGES_REQUESTED** [review]

File: doc/sphinxdoc/models.rst [hunk 8]
```diff
@@ -171,6 +171,97 @@ Usage for embedding extraction:
...
+chime`, `Wind instrument and woodwind instrument`, `Wood`, `Writing`, `Yell`,
+`Zipper (clothing)`
+
+Dataset: FSD50K.
+
+Output: class activations or embeddings.
+
+Models:
+
+* ``fsd-sinet-vgg41-tlpf-ibp`` [`weights <https://essentia.upf.edu/models/audio-event-recognition/fsd-sinet/fsd-sinet-vgg41-tlpf-ibp-1.pb>`_, `metadata <https://essentia.upf.edu/models/audio-event-recognition/fsd-sinet/fsd-sinet-vgg41-tlpf-ibp-1.json>`_]
```
[the remark this answers is not in the document]
Comment (palonso) [reply doc/sphinxdoc/models.rst:228]: that's a good idea. I'll work on that

File: src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp [hunk 9]
```diff
@@ -159,14 +159,14 @@ const char* TensorflowPredictFSDSINet::description = DOC(
...
-  "Note II: The FSD-SINet models were trained on normalized audio clips. "
+  "Note: The FSD-SINet models were trained on normalized audio clips. "
   "Clip-level normalization is only implemented in standard mode since in streaming there is no access to the entire audio clip. "
-  "In this case, it is the responsibility of the user to feed the algorithm with audio featuring a healthy dynamic range.\n\n"
+  "In the streaming case, the user is responsible for controlling the dynamic range of the input signal. "
+  "Ideally, the signal should be zero-mean (no DC) and normalized to the full dynamic range (-1, 1).\n\n"
   "References:\n"
   "  [1] Fonseca, E., Ferraro, A., & Serra, X. (2021). Improving sound event classification by increasing shift invariance in convolutional neural networks. arXiv preprint arXiv:2107.00623.\n"
-  "  [2] https://github.com/edufonseca/shift_sec"
+  m"  [2] https://github.com/edufonseca/shift_sec"
```
Comment (dbogdanov) [inline src/algorithms/machinelearning/tensorflowpredictfsdsinet.cpp:168]: Misspell char `m`.

**Review by dbogdanov — CHANGES_REQUESTED** [review]

**Pull request merged**